politeiad/backendv2/tstorebe/store/mysql/encrypt.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2020-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/decred/politeia/util"
+	"github.com/marcopeereboom/sbox"
+)
+
+func (s *mysql) encrypt(ctx context.Context, tx *sql.Tx, data []byte) ([]byte, error) {
+	// Create a new nonce value
+	_, err := tx.ExecContext(ctx, "INSERT INTO nonce () VALUES ();")
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the nonce value that was just created
+	rows, err := tx.QueryContext(ctx, "SELECT LAST_INSERT_ID();")
+	if err != nil {
+		return nil, fmt.Errorf("query: %v", err)
+	}
+	defer rows.Close()
+
+	var i int64
+	for rows.Next() {
+		if i > 0 {
+			// There should only ever be one row returned. Something is
+			// wrong if we've already scanned the nonce and its still
+			// scanning rows.
+			return nil, fmt.Errorf("multiple rows returned for nonce")
+		}
+		err = rows.Scan(&i)
+		if err != nil {
+			return nil, fmt.Errorf("scan: %v", err)
+		}
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, fmt.Errorf("next: %v", err)
+	}
+	if i == 0 {
+		return nil, fmt.Errorf("invalid 0 nonce")
+	}
+
+	log.Tracef("Encrypting with nonce: %v", i)
+
+	// Prepare nonce
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(i))
+	n, err := sbox.NewNonceFromBytes(b)
+	if err != nil {
+		return nil, err
+	}
+	nonce := n.Current()
+
+	// Encrypt blob
+	s.keyMtx.RLock()
+	defer s.keyMtx.RUnlock()
+
+	return sbox.EncryptN(0, s.key, nonce, data)
+}
+
+func (s *mysql) decrypt(data []byte) ([]byte, uint32, error) {
+	s.keyMtx.RLock()
+	defer s.keyMtx.RUnlock()
+
+	return sbox.Decrypt(s.key, data)
+}
+
+func (s *mysql) zeroKey() {
+	s.keyMtx.Lock()
+	defer s.keyMtx.Unlock()
+
+	util.Zero(s.key[:])
```

Comment:
If we are doing this song and dance you need to check for key == nil everywhere.

Reply:
The key is not an optional value.  It is not possible to initialize a mysql kv store with the key set to nil.  If the key is nil then something is wrong and panicing is the correct behavior.  You agree?

Reply:
Not really; its set to nil on exit.

I would much rather load the encrypted key (password provided by env) from the KV store directly into the structure and then pass a pointer to the various crypto functions.

```
type kvcontext struct {
  key [32]byte
}

func (k *kvcontext) get(x) ([]byte,error) {
  encryptedBlob,err := k.get(x)
  if err != nil {
    return nil, err
  }
  return decrypt(encryptedBlob, &k.key)
}
```

Reply:
Done

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/store/mysql/mysql.go#L62-L88
https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/store/mysql/encrypt.go#L110-L145

Reply:
I don't love the `emptyNonce` bit. Just return a pointer so that we can return nil.

https://github.com/lukebp/politeia/blob/tlog/politeiad/backendv2/tstorebe/store/mysql/mysql.go#L65-L73
Don't do that. If something goes wrong the map may be half encrypted and the other half isn't. Create an array or a new map to stuff encrypted junk in instead.